### PR TITLE
[PTD] export stream id in process group nccl

### PR DESCRIPTION
**Pull request by xunnanxu**
Summary:
In our pipeline parallel use cases, we need to stash multiple activations in CPU memory.
We don't want to just use existing AC support as that would lead to recomputing the activation which is sometimes cost prohibitive.

To better time the overlap, we want to do roughly this:
* P2P comm in stream X during fwd pass
* As soon as send/recv finishes, we want to offload the activation to CPU mem (pre-pinned) as a callback in a separate copy stream Y (in other words, we schedule `Y.wait_stream(X)`)
* In bwd pass, before actual bwd starts we want to make sure the activation is copied back to GPU (either current waits for Y, or X waits for Y, whichever suits).

Today, NCCL pg does not expose the stream while it does hold onto A stream from pool once the NCCL communicator is established. This exposes that.

Test Plan: tested using an internal lib for now

Differential Revision: D57496237

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 1]
```diff
@@ -638,6 +638,11 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
   void performNocolorSplit(at::Device device);
 
+  c10::StreamId getStreamId(
+      OpType opType,
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:642]: I'm trying to understand the arguments in this new API, wondering why do we need `opType, peer, device`? IMO given that pg already associate with/hold a stream internally, this API should simply just return that stream?

Comment (xunnanxu) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:642]: > pg already associate with/hold a stream internally

Not sure if I missed something but I believe it doesn't work that way today?
IIUC, the nccl communicator is lazily created, bound to the pairs (either collective or p2p) and cached (reused for future comms using the same pair as the current).

So to your question, the API reflects this property (see the cpp counterpart).

**Comment by wconstab** [comment]
discussed offline, seems the intention was to get a way to run operations on a user stream that wait for particular comm operations to finish.  There is actually no need to expose the coms stream, since it is already implicitly exposed via the work object.

for example

```
work = dist.send()
with user_stream:
   work.wait()
   additional_user_ops()
```
should cause 'user_stream' to wait for the send operation to finish on its own stream, and then run any additional 'additional_user_ops` scheduled on user_stream after it

**Comment by linhu-nv** [comment]
hi @xunnanxu @wconstab , I was doing some communication-computation overlapping in a moe all2all communication pattern, and also try to use distributed API in pytorch. And I do think that expose stream id for distributed collective ops in torch is necessary (or allow users to determine with stream to use in the collective ops). For example, I have the codes like this:

1. all_gather meta info of token dispatching
all_gather_work = dist.all_gather_into_tensor()
all_gather_work.wait()
	
2. Process meta info of global router, get input_split_sizes and output_split_sizes for all2all comm
input_split_sizes, output_split_sizes =  		
    function_xxx(num_tokens_per_expert_per_rank) 

3.copy split sizes to host memory
input_split_sizes_list = input_split_sizes.to("cpu").tolist()  	
output_split_sizes_list = output_split_sizes.to("cpu").tolist() 
	
4. All2all-v 
handle = dist.all_to_all_single(…, output_split_sizes_list, input_split_sizes_list, …)
handle.wait()

due to computation dependency, the timeline of cpu and gpu looks like this:

截屏2025-01-29 13 18 36

we can see that cpu and gpu need many syncs, and it's hard to separate gpu behaviors with GPU behaviors, so making it difficult to overlap communication with other computations. But if collective communications can bind to user defined streams, it would be much different: we can launch compute/comm ops with dependencies into a same stream, so they can be totally deteched with cpus. 

Your comments or sugguestions will be much appreciated. Thanks.

**Comment by linhu-nv** [comment]
also cc @kwen2501 @mrshenli for help, thanks!

**Pull request closed without merging**